File: android/src/main/java/com/mcodex/nitroarchive/domain/ArchiveError.kt
package com.mcodex.nitroarchive.domain

sealed class ArchiveDomainError(
    val code: String,
    message: String,
    cause: Throwable? = null
) : Exception(message, cause) {

    class InvalidArgument(detail: String) : ArchiveDomainError("E_INVALID_ARGUMENT", "Invalid argument: $detail")
    class InvalidState(detail: String) : ArchiveDomainError("E_INVALID_STATE", "Invalid state: $detail")
    class Disposed : ArchiveDomainError("E_DISPOSED", "Object has been disposed")
    class InvalidArchive(detail: String) : ArchiveDomainError("E_INVALID_ARCHIVE", "Invalid archive: $detail")
    class UnsupportedFormat(detail: String) : ArchiveDomainError("E_UNSUPPORTED_FORMAT", "Unsupported format: $detail")
    class UnsupportedCompression(detail: String) : ArchiveDomainError("E_UNSUPPORTED_COMPRESSION", "Unsupported compression: $detail")
    class UnsupportedEncryption : ArchiveDomainError("E_UNSUPPORTED_ENCRYPTION", "Encryption is not supported")
    class EncryptedArchive : ArchiveDomainError("E_ENCRYPTED_ARCHIVE", "Archive is encrypted")
    class PasswordRequired : ArchiveDomainError("E_PASSWORD_REQUIRED", "Password is required")
    class BadPassword : ArchiveDomainError("E_BAD_PASSWORD", "Incorrect password")
    class EntryNotFound(detail: String) : ArchiveDomainError("E_ENTRY_NOT_FOUND", "Entry not found: $detail")
    class ChecksumMismatch(detail: String) : ArchiveDomainError("E_CHECKSUM_MISMATCH", "Checksum mismatch: $detail")
    class TruncatedArchive : ArchiveDomainError("E_TRUNCATED_ARCHIVE", "Archive is truncated")
    class PathTraversal(detail: String) : ArchiveDomainError("E_PATH_TRAVERSAL", "Path traversal: $detail")
    class AbsolutePath(detail: String) : ArchiveDomainError("E_ABSOLUTE_PATH", "Absolute path rejected: $detail")
    class UnsafeSymlink(detail: String) : ArchiveDomainError("E_UNSAFE_SYMLINK", "Unsafe symlink: $detail")
    class SpecialFile(detail: String) : ArchiveDomainError("E_SPECIAL_FILE", "Special file rejected: $detail")
    class DuplicateEntry(detail: String) : ArchiveDomainError("E_DUPLICATE_ENTRY", "Duplicate entry: $detail")
    class ExtractionLimitExceeded(limit: String, value: String) : ArchiveDomainError("E_EXTRACTION_LIMIT_EXCEEDED", "Extraction limit exceeded: $limit=$value")
    class ArchiveTooLarge : ArchiveDomainError("E_ARCHIVE_TOO_LARGE", "Archive is too large")
    class BufferLimitExceeded : ArchiveDomainError("E_BUFFER_LIMIT_EXCEEDED", "Buffer limit exceeded")
    class InsufficientStorage : ArchiveDomainError("E_INSUFFICIENT_STORAGE", "Insufficient storage")
    class DestinationExists(detail: String) : ArchiveDomainError("E_DESTINATION_EXISTS", "Destination exists: $detail")
    class PermissionDenied(detail: String) : ArchiveDomainError("E_PERMISSION_DENIED", "Permission denied: $detail")
    class PermissionRequired(detail: String) : ArchiveDomainError("E_PERMISSION_REQUIRED", "Permission required: $detail")
    class SecurityScopeRequired : ArchiveDomainError("E_SECURITY_SCOPE_REQUIRED", "Security scope required")
    class BookmarkStale : ArchiveDomainError("E_BOOKMARK_STALE", "Bookmark is stale")
    class ReadOnly : ArchiveDomainError("E_READ_ONLY", "Resource is read-only")
    class UriPermissionRevoked : ArchiveDomainError("E_URI_PERMISSION_REVOKED", "URI permission revoked")
    class UriNotSeekable : ArchiveDomainError("E_URI_NOT_SEEKABLE", "URI is not seekable")
    class ProviderUnsupported : ArchiveDomainError("E_PROVIDER_UNSUPPORTED", "Provider is unsupported")
    class FileNotAvailable(detail: String) : ArchiveDomainError("E_FILE_NOT_AVAILABLE", "File not available: $detail")
    class OperationCancelled : ArchiveDomainError("E_OPERATION_CANCELLED", "Operation cancelled")
    class IoError(detail: String) : ArchiveDomainError("E_IO", "I/O error: $detail")
    class Internal(detail: String) : ArchiveDomainError("E_INTERNAL", "Internal error: $detail")
}

File: ios/Domain/ArchiveError.swift
import Foundation

enum ArchiveDomainError: Error, CustomNSError, Equatable {
  case invalidArgument(String)
  case invalidState(String)
  case disposed
  case invalidArchive(String)
  case unsupportedFormat(String)
  case unsupportedCompression(String)
  case unsupportedEncryption
  case encryptedArchive
  case passwordRequired
  case badPassword
  case entryNotFound(String)
  case checksumMismatch(String)
  case truncatedArchive
  case pathTraversal(String)
  case absolutePath(String)
  case unsafeSymlink(String)
  case specialFile(String)
  case duplicateEntry(String)
  case extractionLimitExceeded(limit: String, value: String)
  case archiveTooLarge
  case bufferLimitExceeded
  case insufficientStorage
  case destinationExists(String)
  case permissionDenied(String)
  case permissionRequired(String)
  case securityScopeRequired
  case bookmarkStale
  case readOnly
  case uriPermissionRevoked
  case uriNotSeekable
  case providerUnsupported
  case fileNotAvailable(String)
  case operationCancelled
  case ioError(String)
  case `internal`(String)

  var code: String {
    switch self {
    case .invalidArgument: return "E_INVALID_ARGUMENT"
    case .invalidState: return "E_INVALID_STATE"
    case .disposed: return "E_DISPOSED"
    case .invalidArchive: return "E_INVALID_ARCHIVE"
    case .unsupportedFormat: return "E_UNSUPPORTED_FORMAT"
    case .unsupportedCompression: return "E_UNSUPPORTED_COMPRESSION"
    case .unsupportedEncryption: return "E_UNSUPPORTED_ENCRYPTION"
    case .encryptedArchive: return "E_ENCRYPTED_ARCHIVE"
    case .passwordRequired: return "E_PASSWORD_REQUIRED"
    case .badPassword: return "E_BAD_PASSWORD"
    case .entryNotFound: return "E_ENTRY_NOT_FOUND"
    case .checksumMismatch: return "E_CHECKSUM_MISMATCH"
    case .truncatedArchive: return "E_TRUNCATED_ARCHIVE"
    case .pathTraversal: return "E_PATH_TRAVERSAL"
    case .absolutePath: return "E_ABSOLUTE_PATH"
    case .unsafeSymlink: return "E_UNSAFE_SYMLINK"
    case .specialFile: return "E_SPECIAL_FILE"
    case .duplicateEntry: return "E_DUPLICATE_ENTRY"
    case .extractionLimitExceeded: return "E_EXTRACTION_LIMIT_EXCEEDED"
    case .archiveTooLarge: return "E_ARCHIVE_TOO_LARGE"
    case .bufferLimitExceeded: return "E_BUFFER_LIMIT_EXCEEDED"
    case .insufficientStorage: return "E_INSUFFICIENT_STORAGE"
    case .destinationExists: return "E_DESTINATION_EXISTS"
    case .permissionDenied: return "E_PERMISSION_DENIED"
    case .permissionRequired: return "E_PERMISSION_REQUIRED"
    case .securityScopeRequired: return "E_SECURITY_SCOPE_REQUIRED"
    case .bookmarkStale: return "E_BOOKMARK_STALE"
    case .readOnly: return "E_READ_ONLY"
    case .uriPermissionRevoked: return "E_URI_PERMISSION_REVOKED"
    case .uriNotSeekable: return "E_URI_NOT_SEEKABLE"
    case .providerUnsupported: return "E_PROVIDER_UNSUPPORTED"
    case .fileNotAvailable: return "E_FILE_NOT_AVAILABLE"
    case .operationCancelled: return "E_OPERATION_CANCELLED"
    case .ioError: return "E_IO"
    case .internal: return "E_INTERNAL"
    }
  }
}
